packages/bvaughn-architecture-demo/src/contexts/SelectedFrameContext.tsx

```diff
@@ -0,0 +1,39 @@
+import { FrameId, PauseId } from "@replayio/protocol";
+import { createContext, PropsWithChildren, useCallback, useContext, useMemo } from "react";
+import useLoadedRegions from "../hooks/useRegions";
+import { ReplayClientContext } from "shared/client/ReplayClientContext";
+import { TimelineContext } from "./TimelineContext";
+import { isPointInRegions } from "shared/utils/time";
+import { getPauseForExecutionPoint } from "../suspense/PauseCache";
+
+export interface SelectedFrameContextType {
+  getSelectedFrameId: () => {
+    pauseId: PauseId;
+    frameId: FrameId;
+  } | null;
+}
+
+export const SelectedFrameContext = createContext<SelectedFrameContextType>({
+  getSelectedFrameId: () => null,
+});
+
+export function SelectedFrameContextRoot({ children }: PropsWithChildren<{}>) {
+  const client = useContext(ReplayClientContext);
+  const loadedRegions = useLoadedRegions(client);
+  const { executionPoint } = useContext(TimelineContext);
+
+  const isLoaded = loadedRegions !== null && isPointInRegions(executionPoint, loadedRegions.loaded);
+
+  function getSelectedFrameId() {
+    const pause = isLoaded ? getPauseForExecutionPoint(client, executionPoint) : undefined;
+
+    const pauseId = pause?.pauseId;
+    const frameId = pause?.data.frames?.[0]?.frameId ?? null;
+
+    return pauseId && frameId ? { pauseId, frameId } : null;
+  }
+
+  const context = { getSelectedFrameId };
+
+  return <SelectedFrameContext.Provider value={context}>{children}</SelectedFrameContext.Provider>;
+}
```

Comment:
I don't understand why this is a _context_. Contexts are stateful things– that you can use to set state from one component(s) and share it with another component(s) in a different part of the tree. This doesn't look like it has any state.

Couldn't `getSelectedFrameId` just be a regular hook (`useSelectedFrameId`)?

(I think it's also generally a good idea to memoize context values (with `useMemo` or `useCallback`) so that subscribers don't over-render.)

Reply:
More succinctly, what I'm suggesting is:
* `SelectedFrameContext` has two values: `selectedFrameId: FrameId | null` and `setSelectedFrameId(frameId: FrameId | null)`.
  * New architecture components _read_ `context.selectedFrameId`
* `SelectedFrameContextAdapter` updates `SelectedFrameContext` when Redux values change (like our other adapters do).

Reply:
> I don't understand why this is a context.

I'm using the context as a dependency injection mechanism: the `SelectedFrameContext` is used by components in the architecture demo package, which doesn't have access to the redux store that contains `selectedFrameId`. So in the architecture demo we use the `SelectedFrameContextRoot` from this file but in devtools we use `SelectedFrameContextAdapter`, which provides `selectedFrameId` from the redux store.

> I think it's also generally a good idea to memoize context values

Agreed, I just couldn't figure out how to memoize a *suspending* value.

> what I'm suggesting is: [...]

Sounds good, that would also fix the memoization issue.